subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/services/SymbolSampleAnalysisEnvironment.kt

```diff
@@ -0,0 +1,118 @@
+import org.jetbrains.kotlin.psi.KtBlockExpression
+import org.jetbrains.kotlin.psi.KtDeclarationWithBody
+import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.psi.KtFunction
+
+internal class SymbolSampleAnalysisEnvironmentCreator(
+    private val context: DokkaContext,
+) : SampleAnalysisEnvironmentCreator {
+
+    private val symbolAnalysisPlugin = context.plugin<SymbolsAnalysisPlugin>()
+
+    override fun <T> use(block: SampleAnalysisEnvironment.() -> T): T {
+        return runBlocking(Dispatchers.Default) {
+            SamplesKotlinAnalysis(
+                sourceSets = context.configuration.sourceSets,
+                context = context,
+                projectKotlinAnalysis = symbolAnalysisPlugin.querySingle { kotlinAnalysis }
+            ).use { kotlinAnalysis ->
+                val sampleAnalysis = SymbolSampleAnalysisEnvironment(
+                    kotlinAnalysis = kotlinAnalysis,
+                    dokkaLogger = context.logger
+                )
+                block(sampleAnalysis)
+            }
+        }
+    }
+}
+
+private class SymbolSampleAnalysisEnvironment(
+    private val kotlinAnalysis: KotlinAnalysis,
+    private val dokkaLogger: DokkaLogger,
+) : SampleAnalysisEnvironment {
+
+    override fun resolveSample(sourceSet: DokkaSourceSet, fullyQualifiedLink: String): SampleSnippet? {
+        val psiElement = findPsiElement(sourceSet, fullyQualifiedLink)
+        if (psiElement == null) {
+            dokkaLogger.warn(
+                "Unable to resolve a @sample link: \"$fullyQualifiedLink\". Is it used correctly? " +
+                        "Expecting a link to a reachable (resolvable) top-level Kotlin function."
+            )
+            return null
+        } else if (psiElement.containingFile !is KtFile) {
+            dokkaLogger.warn("Unable to resolve non-Kotlin @sample links: \"$fullyQualifiedLink\"")
+            return null
+        } else if (psiElement !is KtFunction) {
+            dokkaLogger.warn("Unable to process a @sample link: \"$fullyQualifiedLink\". Only function links allowed.")
+            return null
+        }
+
+        val imports = processImports(psiElement)
+        val body = processBody(psiElement)
+
+        return SampleSnippet(imports, body)
+    }
+
+    private fun findPsiElement(sourceSet: DokkaSourceSet, fqLink: String): PsiElement? {
+        val analysisContext = kotlinAnalysis[sourceSet]
+        return analyze(analysisContext.mainModule) {
+            val isRootPackage = !fqLink.contains('.')
+            // supposed because we expect the last element to be a top-level function, not class/class+function
+            val supposedFunctionName = if (isRootPackage) fqLink else fqLink.substringAfterLast(".")
+            val supposedPackageName = if (isRootPackage) "" else fqLink.substringBeforeLast(".")
+
+            getTopLevelCallableSymbols(FqName(supposedPackageName), Name.identifier(supposedFunctionName)).firstOrNull()?.psi
```

Comment:
The called function `getTopLevelCallableSymbols` returns top-level functions only, which is not as useful as what we have in the Descriptors implementation.

Is there a way to get any symbol here by fully qualified name? That is, a Kotlin or a Java class or anything else?

I've added some user-friendly warning log messages that only Kotlin functions are allowed, or that links to Java declarations are not allowed, but I didn't figure out how to do the same with symbols API.

Any ideas / pointers?

Reply:
> There is resolveKDocTextLink.

Had to do a bit of refactoring (I needed to get `KtSymbol` and not `DRI`), but it works, thanks! Commit here: https://github.com/Kotlin/dokka/commit/92aa97ae4727d3d086ef2481f223eb493a8d1103

> But for Kotlin it should resolve only functions (not only top-level), see https://kotlinlang.org/docs/kotlin-doc.html#sample-identifier

Yep, there are test cases for it, and they all pass now, so everything should be correct. I needed to resolve any link (even if it's not a function) to provide a user-friendly warning message.